### PR TITLE
Switch to TypeScript

### PR DESCRIPTION
**Pull request by szmarczak**
TODO after this gets merged:
- https://github.com/apify/apify-js/issues/1008
- fix leaking alpn negotiation

Closes #21

### CODE REVIEW

File: src/agent/transform-headers-agent.ts
Comment (mnmkng) [inline src/agent/transform-headers-agent.ts:17]: I think we can delete the `@param` annotations when we have TS now. The comments might get out of sync with the actual types over time.

File: src/agent/transform-headers-agent.ts
Comment (mnmkng) [inline src/agent/transform-headers-agent.ts:52]: Is this a conflict between ESLint and TS expectations?

File: src/agent/transform-headers-agent.ts
Comment (mnmkng) [inline src/agent/transform-headers-agent.ts:146]: I think we're not using default exports in the packages migrated to TypeScript. Named export would be better.

File: src/hooks/browser-headers.ts
Comment (mnmkng) [inline src/hooks/browser-headers.ts:60]: Can we type this at line 37 where the variable is defined?

File: src/hooks/http2.ts
Comment (mnmkng) [inline src/hooks/http2.ts:7]: Another http2-wrapper typings issue? Fix what?

File: src/hooks/options-validation.ts
Comment (mnmkng) [inline src/hooks/options-validation.ts:6]: The possible options are quite well known, no?

File: src/index.ts
Comment (mnmkng) [inline src/index.ts:1]: ?

File: test/custom-options.test.js
Comment (mnmkng) [inline test/custom-options.test.js:1]: I think we prefer testing source directly, but this can be done later.

File: test/proxy.test.js
Comment (mnmkng) [inline test/proxy.test.js:6]: this is why we don't use default exports. to prevent `.default` requires.

File: tsconfig.json
Comment (mnmkng) [inline tsconfig.json:5]: Please check the configs in https://github.com/apify/apify-storage-local-js and https://github.com/apify/browser-pool

**Review by mnmkng — CHANGES_REQUESTED** [review]
Some nitpicks, but overall I think we need to wait for @B4nan before merging.

**Comment by mnmkng** [comment]
cc @vladfrangu

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:52]: Correct

File: src/hooks/http2.ts
Comment (szmarczak) [reply src/hooks/http2.ts:7]: Type conflict.

File: test/custom-options.test.js
Comment (szmarczak) [reply test/custom-options.test.js:1]: WDYM?

File: src/hooks/options-validation.ts
Comment (szmarczak) [reply src/hooks/options-validation.ts:6]: They aren't known upon validation. We could type `asserts options it T` but it still doesn't matter as this is a hook.

File: test/custom-options.test.js
Comment (mnmkng) [reply test/custom-options.test.js:1]: https://github.com/apify/apify-storage-local-js/blob/f8bae3875e2b6e954ea843764d6a07dd46bba566/test/index.test.ts#L4

File: package.json
Comment (vladfrangu) [reply package.json:26]: So, why are we using this instead of `rimraf`? Other repos use `rimraf` and it works fine 👀 
Plus, while I'm not one to care about dependency trees/sizes, `del-cli` has a tree that's almost 10x the one of `del-cli` 😨

File: package.json
Comment (vladfrangu) [inline package.json:39]: Why is this a script? Isn't it redundant, especially for dev envs?

File: src/agent/transform-headers-agent.ts
Comment (vladfrangu) [inline src/agent/transform-headers-agent.ts:2]:
```suggestion
// @ts-expect-error TODO: Type `header-generator`
```

File: src/agent/transform-headers-agent.ts
Comment (vladfrangu) [inline src/agent/transform-headers-agent.ts:51]: If headers is typed as a Record, why do you need to use `[]`? What's the error?

File: src/hooks/browser-headers.ts
Comment (vladfrangu) [inline src/hooks/browser-headers.ts:3]: Any reason for `import x = require`? Pretty sure we have esModuleInterop

File: src/hooks/browser-headers.ts
Comment (vladfrangu) [inline src/hooks/browser-headers.ts:26]: Can be converted into this:
```suggestion
    return Object.fromEntries(fixedHeaders.values());
```

File: src/index.ts
Comment (vladfrangu) [inline src/index.ts:4]: `import = require` should be changed to `import * as x`

File: src/index.ts
Comment (vladfrangu) [inline src/index.ts:50]: Holy smokes... there's 5 different exports here. Personally, I think it'd be cleaner for a major release to export it as an object `export { got }` and thats all. Would make all of this much cleaner, and compatible with CJS

File: src/scraping-defaults.ts
Comment (vladfrangu) [inline src/scraping-defaults.ts:36]: Could move the export in front of the `const SCRAPING...` and remove this line too, that's a nit tho

File: test/custom-options.test.js
Comment (vladfrangu) [reply test/custom-options.test.js:1]: > [apify/apify-storage-local-js@`f8bae38`/test/index.test.ts#L4](https://github.com/apify/apify-storage-local-js/blob/f8bae3875e2b6e954ea843764d6a07dd46bba566/test/index.test.ts#L4) `import` from `src`, not `dist`. Which would require rewrite of tests to TS, which is why I said that we can do it later.

It would only require changing the jest config to a ts one, like I did in browser-pool/storage-local. Then you can import from src directly (and skip needing to build before testing)

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:51]: Because it's a Record, not an interface. The error is that generic properties must be accessed via the index operator.

File: src/hooks/browser-headers.ts
Comment (szmarczak) [reply src/hooks/browser-headers.ts:3]: That's the only way to access CJS modules AFAIK. `import from` ES modules.

File: src/hooks/browser-headers.ts
Comment (szmarczak) [reply src/hooks/browser-headers.ts:26]: Thanks, didn't know about this :D

File: src/index.ts
Comment (szmarczak) [reply src/index.ts:50]: I agree. /cc @mnmkng

File: package.json
Comment (szmarczak) [reply package.json:39]: tl;dr runs after `npm install` and before `npm publish`

File: src/scraping-defaults.ts
Comment (vladfrangu) [inline src/scraping-defaults.ts:24]: Thats..not what I meant :P

I meant adding the `export` keyword in front of `const SCRAPING...` 😝

Comment (szmarczak) [reply src/scraping-defaults.ts:24]: ah yea I didn't notice it, it's so obv :P

File: src/index.ts
Comment (szmarczak) [reply src/index.ts:4]: `import * as QuickLRU from 'quick-lru';` doesn't work as it also exports a namespace

File: src/agent/transform-headers-agent.ts
Comment (vladfrangu) [reply src/agent/transform-headers-agent.ts:51]: Huh.. I don't have that issue 🤔 
image

File: src/hooks/browser-headers.ts
Comment (szmarczak) [reply src/hooks/browser-headers.ts:3]: With

```
        "allowSyntheticDefaultImports": true,
        "esModuleInterop": true,
```

I'm getting

```
{ default: [class HeaderGenerator] } // import * as HeaderGenerator from 'header-generator';
Uncaught TypeError: HeaderGenerator is not a constructor
```

Comment (vladfrangu) [reply src/hooks/browser-headers.ts:3]: Drop the `* as`, and just `import HeaderGenerator` and it should work

File: package.json
Comment (mnmkng) [reply package.json:39]: That's a good one. In the SDK we've been using the `prepare` script for as long as I remember. Tbh I kinda hate it because it slows installs down, but at the same time, you don't end up running against some old build accidentally.

I think the best solution would be always to run tests against source without the need to build and then we most likely would not need this `prepare` script and we could use `prepublishOnly` and save some seconds on install. So yeah, agree with Vlad, but we first need to convert all tests to TS.

Comment (szmarczak) [reply package.json:39]: > I kinda hate it because it slows installs down, but at the same time

How does it slow down? You have to build the project anyway.

> I think the best solution would be always to run tests against source without the need to build

Do you mean to use `ts-node`?

Comment (vladfrangu) [reply package.json:39]: Look in browser-pool for tests stuffs, you don't have to use ts-node (directly at least)
You'll just need to update the jest config, install the deps, and import from src directly 👍

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:51]: Did you try building?

Comment (vladfrangu) [reply src/agent/transform-headers-agent.ts:51]: VSCode would've errored out if it wasn't valid. I cloned the repo locally and I see what you mean, tho it makes no sense.. You can mark this as resolved

File: src/hooks/browser-headers.ts
Comment (szmarczak) [reply src/hooks/browser-headers.ts]: It worked, thanks :D

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:51]: Let's keep this open. Some magic behavior happening here that needs to be fixed.

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [inline src/agent/transform-headers-agent.ts:73]: @vladfrangu This is exactly slower and is not stricter (nor loose) because `transformedHeaders` has cloned entries.

Comment (vladfrangu) [reply src/agent/transform-headers-agent.ts:73]: A for-in loop cannot be faster than a for-of, and it's far more risky due to prototype pollution, and it's stricter in the context of TS (and it picking just the object's own properties)

Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:73]: 80% slower

<details>

<summary>source</summary>

https://jsbench.me/

#### setup

```js
const o = {
	foo: 'bar',
	hello: 'world',
	connection: 'keep-alive',
	host: 'example.com',
	'content-length': '16',
	'accept-encoding': 'gzip',
	'user-agent': 'lorem impsum'
};
```

#### 1

```js
for (let key in o) {
	o[key];
}
```

#### 2

```js
for (const [key, value] of Object.entries(o)) {}
```

</details>

Comment (vladfrangu) [reply src/agent/transform-headers-agent.ts:73]: Touche. For an ideal world where there's no prototype pollution, for-in might be faster, but not safer. Consider this:
image

Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:73]: > A for-in loop cannot be faster than a for-of

I'd love to see an example :) Haven't really met this case, but I don't negate it.

> it's far more risky due to prototype pollution

No prototype pollution here, the entries are cloned before.

> it's stricter in the context of TS (and it picking just the object's own properties)

You're right. I'll leave this up to you then :)

Comment (vladfrangu) [reply src/agent/transform-headers-agent.ts:73]: > No prototype pollution here, the entries are cloned before.

*We* don't have prototype pollutions 😄 
Other dependencies, user code, etc might tho. Thats where the speed balance shifts 😅

Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:73]: Prototype mutation on the native types should be removed from V8 IMO. It's strongly discouraged to do this. The entire app's behavior can change. Node.js circumvents this via primodals.

Comment (vladfrangu) [reply src/agent/transform-headers-agent.ts:73]: > Prototype mutation on the native types should be removed from V8 IMO. It's strongly discouraged to do this. The entire app's behavior can change. Node.js circumvents this via primodals.

Oh I am aware and I strongly agree, but till that day comes, I think the best compromise we have right now for speed/safety is Object.keys. What do you think?

Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:73]: Dunno. `Object.keys` can be overwritten as well (unlikely tho) 😅 I'm good with anything that works :D

File: package.json
Comment (szmarczak) [reply package.json:26]: > So, why are we using this instead of rimraf?

Dunno. It's just what I use. Will change this.

File: tsconfig.json
Comment (szmarczak) [reply tsconfig.json]: So many config files 🥶

File: src/agent/transform-headers-agent.ts
Comment (vladfrangu) [reply src/agent/transform-headers-agent.ts:73]: CC @mnmkng thoughts? ^

File: tsconfig.json
Comment (vladfrangu) [reply tsconfig.json]: We could make this a module, like the eslint config! Just need to do it 👍

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:51]: Fixed in #39

File: src/index.ts
Comment (mnmkng) [reply src/index.ts:50]: I have only a faint idea how this ESM vs CJS vs TS build arguments exporting magic works. So yeah, I'm definitely for a cleaner API if that can be done. We'll have to address it in the SDK, if the imports will be different, but that's fine.

File: src/agent/transform-headers-agent.ts
Comment (mnmkng) [reply src/agent/transform-headers-agent.ts:73]: I tried to find a nice article I read, but it said something like this: The lengthiest and most heated discussions in code reviews are spent on discussing theoretical problems and code preferences of individual developers 😄 

I personally don't use `for in`. Not because I think it's always bad, but because it makes people argue and it's just not worth arguing about. We have about 15 other ways of writing a loop which do not promote heated discussions ❤️

The performance here is irrelevant really. For each request that will take on average hundreds of milliseconds, we don't really care if it the client spends 1 microsecond or 2 microseconds in a loop.

So let's just keep whatever loop is in there now 🤝

File: tsconfig.json
Comment (vladfrangu) [reply tsconfig.json]: Update: it's been done! @apify/tsconfig is available, check out browser-pool for how to use it 😋. If you want, I can push it for you

File: test/browser-headers.test.ts
Comment (B4nan) [inline test/browser-headers.test.ts:15]: seems like if you type this as just `Options` you will safe yourself from a lot of type casting

File: test/custom-options.test.ts
Comment (B4nan) [inline test/custom-options.test.ts:3]: i believe the linter errors should go away too when you update `@apify/eslint-config-ts` to latest

File: test/agent.test.ts
Comment (B4nan) [inline test/agent.test.ts:84]: maybe the `WrappedAgent` should extend/implement `Agent` interface?

**Review by B4nan — CHANGES_REQUESTED** [review]
left some comments/questions, but looks good in general

maybe we should PR the `@types/node` if they are not correct, or override them locally

File: src/index.ts
Comment (vladfrangu) [reply src/index.ts:50]: I think it'd be best for everyone if we just `export { got }`. That's a major bump for this module, and a patch bump for modules using this module, as well as keeping code consistent for CJS, ESM and TS 👍

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:94]: We can actually remove this but this is breaking for people who pass an Agent based on `agent-base` explicitly. I'll check if `HttpsProxyAgent` exports an options interface.

File: src/hooks/custom-options.ts
Comment (szmarczak) [reply src/hooks/custom-options.ts:8]: I had previously some issues that TypeScript required all generic properties to be accessed with the brackets but it's solved now. I'll fix this to use the dot notation instead.

File: src/agent/transform-headers-agent.ts
Comment (szmarczak) [reply src/agent/transform-headers-agent.ts:94]: Actually we don't use `agent-base` anymore so it has to be typed `as any`.

File: test/tsconfig.json
Comment (vladfrangu) [inline test/tsconfig.json:15]: Don't forget to update this 😋

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: test/tsconfig.json
Comment (szmarczak) [reply test/tsconfig.json]: What about now?

File: src/hooks/proxy.ts
Comment (B4nan) [inline src/hooks/proxy.ts:20]: this definitely looks less out of space then the previous version 👍

**Review by B4nan — CHANGES_REQUESTED** [review]

File: test/tsconfig.json
Comment (vladfrangu) [reply test/tsconfig.json]: That works! 👍

File: package.json
Comment (vladfrangu) [reply package.json:50]: When you run the script through `npm run` / `yarn`, it'll run the binaries from node_modules, so you don't need to put npx in front on them

File: tsconfig.json
Comment (vladfrangu) [reply tsconfig.json:5]: That's actually a redundant line, as `@apify/tsconfig` already has that set to it, it should be removed

File: CHANGELOG.md
Comment (vladfrangu) [reply CHANGELOG.md:3]: Probably the exports will be the biggest breaking change 👀

File: package.json
Comment (szmarczak) [reply package.json:50]: I don't have `eslint` installed globally and it works. Am I missing something?

Comment (B4nan) [reply package.json:50]: I guess something changed in npm 7 then, as I remember things like this were causing CI failures historically.

**Comment by szmarczak** [comment]
> this are still a lot of those FIXME type comments, do you plan to resolve all of them before merging the PR? if not, we should be probably more specific instead of just saying FIXME, at least in the actual code base (it's fine'ish in tests)

Fixed the two in `index.ts` and `http2.ts`.

The ones in `main.test.ts` are unfixable atm, see https://github.com/sindresorhus/got/issues/1117

**Comment by szmarczak** [comment]
@B4nan Friendly ping :)

**Review by B4nan — APPROVED** [review]

**Pull request merged**
